src/prng/chacha.rs

```diff
@@ -292,10 +248,96 @@ impl SeedableRng for ChaChaCore {
 
 impl CryptoRng for ChaChaCore {}
 
+/// A builder type allowing extra parameterisation of ChaCha.
+///
+/// Construct using `SeedableRng` as normal.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::{Rng, SeedableRng};
+/// use rand::prng::chacha::{ChaChaRng, ChaChaBuilder};
+///
+/// let mut rng: ChaChaRng =
+///     ChaChaBuilder::from_seed([0u8; 32])
+///         .set_rounds(12)
+///         .set_stream(123456)
+///         .into();
+/// rng.gen::<u32>();
+/// ```
+#[derive(Clone, Debug)]
+pub struct ChaChaBuilder {
+    core: ChaChaCore,
+}
+
+impl ChaChaBuilder {
+    /// Set the number of rounds.
+    ///
+    /// The default variant is ChaCha20, using 20 rounds. Published variants
+    /// use 8 or 12 rounds, where 8 is the smallest number of rounds for which
+    /// ChaCha has not been broken, with more rounds improving security at the
+    /// cost of speed.
+    ///
+    /// The number of rounds must be even.
+    pub fn set_rounds(mut self, rounds: u32) -> Self {
+        assert!(rounds & 1 == 0);
+        self.core.rounds = rounds;
+        self
+    }
+    
+    /// Set a nonce from a byte slice. The length of the slice must correspond
+    /// to the length of the nonce set during construction (which by default is
+    /// zero length).
+    pub fn set_nonce(mut self, nonce: [u8; 8]) -> Self {
+        let mut nonce_le = [0u32; 2];
+        le::read_u32_into(&nonce, &mut nonce_le);
+        self.core.state[14] = nonce_le[0];
+        self.core.state[15] = nonce_le[1];
+        self
+    }
+
+    /// Set the nonce from a stream number.
+    ///
+    /// This simply reinterprets the nonce region as a stream code.
+    pub fn set_stream(mut self, stream: u64) -> Self {
```

Comment:
I was slightly confused about the use of this, but :+1:.

Reply:
Would it be useful with parallel iterators perhaps?

Reply:
Yes, but I suppose there are more situations.